Sources/ServiceKit/Container/Container.swift

```diff
@@ -1,41 +1,41 @@
-import NIO
+@_exported import protocol NIO.EventLoop
+@_exported import class NIO.EventLoopFuture
 
-/// `Container`s are used to create instances of services that your application needs in a configurable way.
-///
-///     let client = try container.make(Client.self)
-///
-/// Each `Container` has `Config`, `Environment`, and `Services`. It uses this information to dynamically provide
-/// services based on your configuration and environment.
-///
-///     switch env {
-///     case .production: config.prefer(ProductionLogger.self, for: Logger.self)
-///     default: config.prefer(DebugLogger.self, for: Logger.self)
-///     }
-///
-///     ...
-///
-///     let logger = try container.make(Logger.self) // changes based on environment
-///
-/// Containers are also `Worker`s, meaning they have a reference to an `EventLoop`.
-///
-///     print(container.eventLoop)
-///
-/// - warning: You should never use services created from a `Container` on _another_ `Container`'s `EventLoop`.
-public protocol Container: class {
+public final class Container {
+    public static func boot(env: Environment = .development, services: Services, on eventLoop: EventLoop) -> EventLoopFuture<Container> {
+        let container = Container(env: env, services: services, on: eventLoop)
+        return container.willBoot()
+            .flatMap { container.didBoot() }
+            .map { container }
+    }
+    
     /// Service `Environment` (e.g., production, dev). Use this to dynamically swap services based on environment.
-    var environment: Environment { get }
-
+    public let env: Environment
+    
     /// Available services. This struct contains all of this `Container`'s available service implementations.
-    var services: Services { get }
+    public let services: Services
     
-    /// Stores cached singleton services.
-    var cache: ServiceCache { get set }
+    /// All `Provider`s that have been registered to this `Container`'s `Services`.
+    public var providers: [Provider] {
+        return self.services.providers
+    }
     
     /// This container's event loop.
-    var eventLoop: EventLoop { get }
-}
-
-extension Container {
+    public let eventLoop: EventLoop
+    
+    /// Stores cached singleton services.
+    private var cache: ServiceCache
+    
+    private var didShutdown: Bool
```

Comment:
I'm not 100% in love with this naming. Can't really think of something better either, but maybe `isActive` and have it on `true` by default?

Reply:
I see your point. I like the sound of `isActive`, but I also like that `didShutdown` makes incredibly clear what the variable tracks (it's set to `true` after `shutdown` is called).